system/include/emscripten/fetch.h

```diff
@@ -184,6 +185,22 @@ EMSCRIPTEN_RESULT emscripten_fetch_wait(emscripten_fetch_t *fetch, double timeou
 // onerror() handler will be called in the calling thread before this function returns.
 EMSCRIPTEN_RESULT emscripten_fetch_close(emscripten_fetch_t *fetch);
 
+// Gets the size (in bytes) of the response headers as plain text.
+// Will only return a sensible result if the ready state is larger than or equal to 2.
+size_t emscripten_fetch_get_response_headers_length(emscripten_fetch_t *fetch);
+
+// Gets the response headers as plain text. dstSizeBytes should be headers_length + 1 (for the null terminator).
+size_t emscripten_fetch_get_response_headers(emscripten_fetch_t *fetch, char *dst, size_t dstSizeBytes);
+
+// Converts the plain text headers into an array of strings. This array takes the form
+// {"key1", "value1", "key2", "value2", "key3", "value3", ..., 0 }; Note especially that the array
+// is terminated with a null pointer.
+char **emscripten_fetch_unpack_response_headers(const char *headersString);
+
+// This frees the memory used by the array of headers. Call this when finished with the data returned
+// by emscripten_fetch_unpack_response_headers.
+void emscripten_fetch_free_unpacked_response_headers(char **unpackedHeaders);
```

Comment:
Should this be called inside emscripten_fetch_close?

Reply:
Could do for safety. I'm not sure if one might want to keep the response headers separate and alive after closing the fetch.

Reply:
The description of emscripten_fetch_close is `Closes a finished or an executing fetch operation and frees up all memory`.
I think either emscripten_fetch_close should call emscripten_fetch_free_unpacked_response_headers, or code comments/documentation need to be updated to let people know you need to call both emscripten_fetch_free_unpacked_response_headers and emscripten_fetch_close to free all the data.
As it is I think it would be very easy for developers to create a memory leak.

Personally I think the first option is better, and if someone wants the headers they just copy the data. Or more likely parse it into a more workable format like a vector of std::strings.

Reply:
I also agree the first option seems better.

Reply:
No, `emscripten_fetch_close` should not call `emscripten_fetch_free_unpacked_response_headers`, because it would not know what to free - the memory management for unpacked response headers fields is separate from the fetch data struct. Making `emscripten_fetch_close` call `emscripten_fetch_free_unpacked_response_headers` would mandate that the fetch structure would have to track all instances of allocated unpacked response headers, and it would add that dependency from core fetch functionality to the logic of unpacking response headers, increasing code size that was desirable to avoid (as per https://github.com/emscripten-core/emscripten/pull/8486#issuecomment-489047322).

Think of the role of `emscripten_fetch_unpack_response_headers` & `emscripten_fetch_free_unpacked_response_headers` to be stateless helper utility functions that offer string splitting functionality on behalf of the user (which they can also do themselves if they like)

Reply:
Oh, for some reason I thought that the `emscripten_fetch_t` struct contained `unpackedHeaders`. Yeah juj is right.